Instant Search: Improve product result format layout

jsnmoon opened the pull request:
#### Changes proposed in this Pull Request:
* Adds a 5-column layout to the product result format at large breakpoints.

<img width="1355" alt="Screen Shot 2021-03-24 at 5 48 48 PM" src="https://user-images.githubusercontent.com/4044428/112398200-3f1b4900-8cc9-11eb-93be-3faf7cb4cfd8.png">


* Adds a placeholder visual if the result lacks a corresponding image.

<img width="1355" alt="Screen Shot 2021-03-24 at 5 49 00 PM" src="https://user-images.githubusercontent.com/4044428/112398212-45a9c080-8cc9-11eb-82e6-d49f5ee8987e.png">

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Apply this change to your site with a Jetpack Search subscription.
* Perform a site search; append `result_format=product` query string to force the product result format layout. Ensure that the results look reasonable on all breakpoints (e.g. desktop, tablet, mobile).

Commit:
Add XL breaks

Commit:
Add a 4 column layout on XL break

Commit:
Add a gridicon placeholder for results w/o images

Commit:
Add dark theme support, fix image spacing issue

bluefuton commented:
The 4-per-row works well on bigger screens 👍 I like it!

It would be good if the placeholder image could match the height of the row, so the product titles lined up. It's like that in your example screenshot, but on my test site:

<img width="1127" alt="Screen Shot 2021-03-05 at 14 11 35" src="https://user-images.githubusercontent.com/17325/110052756-f4d73400-7dbc-11eb-8f83-afc433694543.png">

jsnmoon commented:
> <img width="988" alt="Screen Shot 2021-02-17 at 4 07 11 PM" src="https://user-images.githubusercontent.com/4044428/108279676-34a1e880-713a-11eb-8574-03b98727f519.png">

> <img width="1127" alt="Screen Shot 2021-03-05 at 14 11 35" src="https://user-images.githubusercontent.com/17325/110052756-f4d73400-7dbc-11eb-8f83-afc433694543.png">

@jeffgolenski: Could you recommend a better placeholder design than the one already implemented above? The landscape ratio of the image gridicon doesn't seem to lend itself to the grid layout.

bluefuton reviewed (dismissed):
I think the image placeholder works well enough for the moment :) Just two small things to fix:

- Line up titles for products with images and products with placeholders

<img width="895" alt="Screen Shot 2021-03-10 at 16 15 25" src="https://user-images.githubusercontent.com/17325/110571125-f1c4b500-81bb-11eb-961d-19f59af4ace8.png">

- Update the SVG title for the placeholder (isn't correct in this case):

<img width="363" alt="Screen Shot 2021-03-10 at 16 13 17" src="https://user-images.githubusercontent.com/17325/110571033-cb067e80-81bb-11eb-98dc-bff55fa28c94.png">

Commit:
Replaces default SVG title with appropriate copy

Commit:
Consolidate image and placeholder styles

bluefuton reviewed (commented):
I think you mentioned this earlier today - getting some inconsistent image heights (Photon is turned off):

<img width="913" alt="Screen Shot 2021-03-12 at 16 14 09" src="https://user-images.githubusercontent.com/17325/110886880-36805580-834e-11eb-90ce-b3699a277dfd.png">

Titles are now lining up when images are big enough, but perhaps need a tiny bit more top margin on the title? (cc @jeffgolenski)

IE11 has some row height issues:

<img width="897" alt="Screen Shot 2021-03-12 at 16 14 44" src="https://user-images.githubusercontent.com/17325/110886960-5283f700-834e-11eb-8fc9-959d9ad62c7e.png">

Commit:
Try using a static layout

Commit:
Add changelog

jsnmoon commented:
@bluefuton: I went ahead and changed to a static layout for the product grid. This means that images won't automatically resize to fit the grid, but it guarantees height/width for each image.

Commit:
Remove redundant (and misplaced) style

Commit:
Rename style variable

Commit:
Remove padding from mark elements

bluefuton commented:
One possible solution for IE object-fit: https://www.stevefenton.co.uk/2019/09/fixing-css-object-fit-for-internet-explorer/

It'd be nice if we could use more of the available horizontal space (particularly on bigger screens) but this is much more solid across browsers 👍

<img width="1087" alt="Screen Shot 2021-03-15 at 14 34 37" src="https://user-images.githubusercontent.com/17325/111092760-dc74cf80-859b-11eb-858b-620a17906d16.png">

Commit:
Fix 2021 theme conflict with close button

Commit:
Simplify/fix sidebar modal placement

Commit:
Add dark theme support for close button border fix

Commit:
Reduce product grid width

jsnmoon commented:
Further reduced the grid width to increase content density. I'll punt the object-fit fix for IE11 to our image component enhancement PR.

bluefuton commented:
This is how it currently displays on my usual browser window size:

<img width="1106" alt="Screen Shot 2021-03-17 at 16 04 32" src="https://user-images.githubusercontent.com/17325/111408794-a9b50d80-873a-11eb-8f07-85f0f47a2ae0.png">

Is there an opportunity to use a more flexible grid? (Like https://css-tricks.com/designing-a-product-page-layout-with-flexbox/ ?) Are we constrained by image size?

Commit:
Restore dynamic layout using aspect-ratio rule

jsnmoon commented:
I was able to restore a dynamic flexbox-based layout using the [`aspect-ratio` style rule](https://caniuse.com/mdn-css_properties_aspect-ratio), which appears to be decently supported by most browsers.

It doesn't work with IE11, though.

bluefuton commented:
Thanks @jsnmoon! Is it something other than `aspect-ratio` causing the wonky line height in IE11? I see `aspect-ratio` isn't well supported in current Safari or Firefox either but the product rows look consistent there.

<img width="1278" alt="Screen Shot 2021-03-19 at 12 02 47" src="https://user-images.githubusercontent.com/17325/111708697-341c7f00-88ab-11eb-8bbb-baf27d2e5a86.png">

Commit:
Use aspect-ratio CSS trick

jsnmoon commented:
> Is it something other than `aspect-ratio` causing the wonky line height in IE11? I see `aspect-ratio` isn't well supported in current Safari or Firefox either but the product rows look consistent there.

I decided against using `aspect-ratio` due to its poor support. In its place, I used [a CSS hack to implement a constant aspect ratio box](https://css-tricks.com/aspect-ratio-boxes/).

This solution appears to work reliably for Chrome, Firefox, Safari, and IE11.

Commit:
Add full height/width for placeholder icon

Commit:
Add gridicons-block

Commit:
Improve placeholder visuals, component hierarchy

jsnmoon commented:
Updated placeholder visuals when a search result contains no images:

<img width="1332" alt="Screen Shot 2021-03-19 at 3 30 46 PM" src="https://user-images.githubusercontent.com/4044428/111844192-275a5400-88c8-11eb-8a8c-c905b0b7d2c8.png">

bluefuton commented:
> Updated placeholder visuals when a search result contains no images 

Shall we clamp the description/body text to say, 2 lines for product layout? cc @jeffgolenski 

<img width="305" alt="Screen Shot 2021-03-22 at 15 26 44" src="https://user-images.githubusercontent.com/17325/111932766-06416f80-8b23-11eb-87a0-849abddc59f4.png">

bluefuton commented:
Layout is looking nice across browsers now 🎉

One thing I noticed - the sort selector is unresponsive when positioned on the right. Perhaps hidden behind another element?

<img width="1331" alt="Screen Shot 2021-03-22 at 15 49 00" src="https://user-images.githubusercontent.com/17325/111934427-43f3c780-8b26-11eb-82eb-077e9fc07148.png">

Commit:
Remove unused invocations of sprintf

Commit:
Fix z-index for sort selector

Commit:
Replace content/comments with match snippet

jsnmoon commented:
> Shall we clamp the description/body text to say, 2 lines for product layout? cc @jeffgolenski

I went ahead and replaced the highlights with a keyword match snippet from the design specifications:

<img width="1332" alt="Screen Shot 2021-03-22 at 2 02 41 PM" src="https://user-images.githubusercontent.com/4044428/112051208-469af080-8b17-11eb-8a18-497ade547ef8.png">

> One thing I noticed - the sort selector is unresponsive when positioned on the right. Perhaps hidden behind another element?

This was a z-index issue and has now been fixed.

Commit:
Conditionally render match content

Commit:
Omit gridicon title from match snippet

bluefuton reviewed (commented):
Sort now works again. Thanks! 🎉 

Good call on "keyword found in content" @jsnmoon 👍 It's appearing in bold for me, which draws the eye a bit too much - can we set the font weight to normal?

<img width="295" alt="Screen Shot 2021-03-23 at 11 53 36" src="https://user-images.githubusercontent.com/17325/112068569-0bf48f00-8bcf-11eb-8066-b355b4a29169.png">

One other bit of feedback - it says "keyword found in content" even when I haven't entered a search string, which feels a bit weird. Can we omit it when the search string is empty?

bluefuton commented:
One more request as I'm looking at adding `alt` text for images: is it straightforward to use a `img` element rather than a background-image?

<img width="1080" alt="Screen Shot 2021-03-23 at 12 34 38" src="https://user-images.githubusercontent.com/17325/112071219-4d3b6d80-8bd4-11eb-97e8-9e6a8dec79a0.png">

Commit:
Override themes' mark styling

Commit:
Prevent img alt from being visible on broken images

Commit:
Fix references to searchQuery

Commit:
Use photon img, add alt

Commit:
Show match prompt iff search query exists

jsnmoon commented:
> Good call on "keyword found in content" @jsnmoon 👍 It's appearing in bold for me, which draws the eye a bit too much - can we set the font weight to normal? (@bluefuton)

Good catch, this is likely caused by the theme. 

> One other bit of feedback - it says "keyword found in content" even when I haven't entered a search string, which feels a bit weird. Can we omit it when the search string is empty? (@bluefuton)

I'm slightly against this idea since our API is essentially indicating a keyword match via the result's `highlight` field. I'll go ahead and make the change here for now, but let's prioritize improving the API's `highlight` output in the future (e.g. returning `null` instead of `[""]` for non-matches).

> One more request as I'm looking at adding `alt` text for images: is it straightforward to use a `img` element rather than a background-image? (@bluefuton)

Done! I went ahead and set the `alt` to the result's `title` for now.

jsnmoon commented on projects/plugins/jetpack/modules/search/instant-search/components/search-result-product.scss:
```diff
@@ -1,43 +1,112 @@
 .jetpack-instant-search__search-result-product-img {
-	margin: 0.5em 0.5em 0.5em 0;
-	max-width: 100%;
+	background-size: cover;
+	background-repeat: no-repeat;
+	background-position: center;
```
These styles are no longer needed, will delete soon.

Commit:
Remove unused styles, add more comments

bluefuton reviewed (commented):
Looking so much more polished now! Excited to get this launched :)

I found one display issue with long, tall portrait images obscuring the title (do we need a `max-height`?):

<img width="1086" alt="Screen Shot 2021-03-24 at 12 23 41" src="https://user-images.githubusercontent.com/17325/112231632-ee412b80-8c9b-11eb-97e2-e1d142f481b5.png">

This is the errant image:

![1000071377_070_Zoom_3-Full_Length](https://user-images.githubusercontent.com/17325/112231658-fc8f4780-8c9b-11eb-846b-6d6e4505821b.jpg)

bluefuton commented:
> I'm slightly against this idea since our API is essentially indicating a keyword match via the result's highlight field. I'll go ahead and make the change here for now, but let's prioritize improving the API's highlight output in the future

Created https://github.com/Automattic/jetpack/issues/19272 to keep track of this 👍

Commit:
Add styles to ensure proper img placement

jsnmoon commented:
> I found one display issue with long, tall portrait images obscuring the title (do we need a max-height?):

Looks like this was a regression caused by switching from `div` to `img`. I've fixed it using `height`, `width`, and `object-fit` styles.

jeffgolenski commented:
Hey hey @jsnmoon & @bluefuton this is really starting to look AMAZING. 

I'm running this branch live on midnightshift.photo because I'm reckless. haha

Here's a quick design review so far with a video of various screen sizes. Please ignore anything you're aware of:


https://user-images.githubusercontent.com/214813/112383799-f4e39900-8cc3-11eb-920e-331465a40338.mp4

* Image scaling is strange with the gray block behind the image
* Results that highlight the keyword in the title, don't need to have the "keyword found in content" label
* The sorting is functional and works great—we should just clean up the font and sizing of the box there. Let's make the drop down UI a bit bigger and the font a bit larger
* Can we make "clear" a little larger as well?
* On mobile / small screens, can the sort by occupy the entire width of the screen for easier tapping

* ~0:55 in video: When a search phrase is removed and the field is empty are we just displaying whatever is found in /shop/? In this instance, I like how you've defaulted it to say "showing popular results." Do you think there's a need to remind people to enter a phrase when that field is empty?

Commit:
Increase 'clear' button font size

Commit:
Increase sort select font size, add padding and border radius.

Commit:
Omit match hint if title matches

Commit:
Add 4 column break at medium breakpoint

Commit:
Fix primary area margins at differing breakpoints

Commit:
Split controls from SearchForm for better DOM flow

Commit:
Fix primary margins at differing breakpoints 2

Commit:
Improve controls layout + look & feel

Commit:
Shorten match hint

jsnmoon commented:
> - [x] Image scaling is strange with the gray block behind the image (@jeffgolenski)

Fixed in df604b0e6ab58982609608e03b3ce6165dc9eea4!

> - [x] Results that highlight the keyword in the title, don't need to have the "keyword found in content" label (@jeffgolenski)

Fixed in c1fec7950c4c681720081ea930f729cc14e6c699!

> - [x] The sorting is functional and works great—we should just clean up the font and sizing of the box there. Let's make the drop down UI a bit bigger and the font a bit larger (@jeffgolenski)
> - [x] Can we make "clear" a little larger as well? (@jeffgolenski)
> - [x] On mobile / small screens, can the sort by occupy the entire width of the screen for easier tapping (@jeffgolenski)

Done!

> - [x] When a search phrase is removed and the field is empty are we just displaying whatever is found in /shop/? (@jeffgolenski)

I think so, yes. I'm under the impression that we might be scoring (and thus sorting) the default popular results differently than in `/shop/`.

> Do you think there's a need to remind people to enter a phrase when that field is empty? (@jeffgolenski)

I don't think that's necessary 😊

---

I also made two additional changes:
- Reduced the verbosity of the match hint to say `Matches content`/`Matches comment` instead of `Keyword found in content`/`Keyword found in comment`.
- Reduced the verbosity of the sort label to say `Sort:` instead of `Sort by:`.

Commit:
Adjust sort select position at larger breakpoints

Commit:
Merge branch 'master' of github.com:Automattic/jetpack into fix/improve-search-product-layout

jsnmoon commented:
I refactored `SearchControls` out of the `SearchForm` component to simplify component styling. This can technically impact other result formats, so please test minimal and expanded result formats if you're able.

bluefuton reviewed (approved):
<img src="https://user-images.githubusercontent.com/17325/112410739-929c9e00-8d80-11eb-96e6-c18321896e49.png" width="200" />

Lovely. Tested out well for me on Firefox, Chrome and Edge.

Two minor nits but these don't need to hold up the approval:

*  The "Showing popular results" header now feels a bit close to the sort selector on the smallest breakpoint.

<img width="280" alt="Screen Shot 2021-03-25 at 15 29 47" src="https://user-images.githubusercontent.com/17325/112410495-228e1800-8d80-11eb-8462-fa9a56bca0c3.png">

* A search string that just contains whitespace (" ") will trigger the "matches content" message - should we trim the string before doing that check? e.g. http://example.com/?s=%20&result_format=product. Totally acknowledge that this best fixed in the API ultimately (filed in https://github.com/Automattic/jetpack/issues/19272).

kangzj commented:
Awesome work @jsnmoon 👍 
Just had a spin, there was something which might need some attention. On wider screens like Jeff's, the image could get a little bit too big.

https://d.pr/i/dXe539

jeherve reviewed (approved):
It tested well for me. 🚢

jsnmoon commented:
r223374-wpcom.
